File: hll/autogen/examples/helloWorld/helloworld.c
#ifndef INCLUSION_HELLOWORLD_H
#include "helloworld.h"
#endif

#include "ocr.h" //PRINTF

#include <math.h> //sqrt

const double nota_root = -1e300;

static int equal_double( double x, double y)
{
    double diff = x - y;
    if(diff<0) diff = - diff;
    if(diff > 1e-14) return 0; //Zero for false
    return 1; //1 for true
}

Err_t setupTheParabola(int * io_parabola)
{
    Err_t erri = 0;
    //a -> 2, b -> 1, c -> 0  and y = a*x^2 + b*x + c
    io_parabola[0] = 1;
    io_parabola[1] = 2;
    io_parabola[2] = 1;

    PRINTF("TESTIO> Setup used the parabola (x^2,x,1) = (%d, %d, %d).\n",
            io_parabola[2],
            io_parabola[1],
            io_parabola[0]
            );

    return erri;
}
Err_t solveForRoots(int * in_parabola, double * o_roots)
{
    Err_t erri = 0;
    //a -> 2, b -> 1, c -> 0  and y = a*x^2 + b*x + c
    const double c = in_parabola[0];
    const double b = in_parabola[1];
    const double a = in_parabola[2];

    //See the book "Numerical recipes in C" for more detail.
    const double signb = ((b<0)?(-1):(1));
    const double q = (b + signb*sqrt(b*b - 4*a*c))/2;
    o_roots[0] = q/a;
    o_roots[1] = c/q;

    PRINTF("TESTIO> The Solver for roots %21.14e and %21.14e.\n", o_roots[0], o_roots[0]);

    return erri;
}
Err_t deduplicateRoots(double * io_roots)
{
    PRINTF("TESTIO> Checking for duplicate roots.\n");
    Err_t erri = 0;

    if( equal_double(io_roots[0],io_roots[1]) ){
        io_roots[1] = nota_root;
    }

    return erri;
}
Err_t printRoots(double * io_roots)
{
    Err_t erri = 0;

    if( equal_double(io_roots[1], nota_root)){
        PRINTF("TESTIO> The double root is %21.14e\n", io_roots[0]);
    } else {
        PRINTF("TESTIO> The first and second roots are %21.14e and %21.14e\n", io_roots[0], io_roots[1]);
    }

    return erri;
}

File: hll/autogen/examples/helloWorld/helloWorld.py
import sys
sys.path.append('../../tools')
import ocr_autogen as OA
import copy

# ------------------------------------------------------------------------------
def theMain():

    OA.GBL.insert_debug_code = 0

    G=OA.GraphTool.DiGraph()

    OA.setupGraph(G,"HelloWorld")

    OA.addHeader(G, '#include "helloworld.h" ')  # This will be put at the beginning of the file

    # ----- Data Blocks
    dbk = OA.ocrDataBlock();  dbk.name = 'NULL_GUID'
    dbk_nullGuid = dbk

    dbk = OA.ocrDataBlock();  dbk.name = 'Parabola';   dbk.count=3;     dbk.type='int'
    dbk_parabola = dbk

    dbk=OA.ocrDataBlock();  dbk.name = 'Roots';   dbk.count=2;     dbk.type='double'
    dbk_roots = dbk

    # ----- NODES
    # Always root the starting/root/main EDT at ID=0 = OA.GBL.MAINNODE
    # Always cap all ending EDT by the FinalEDT at ID=1 = OA.GBL.FINALNODE
    nc = OA.GBL.MAINNODE;    taskName = "mainEdt"; OA.graphAddNode(G,nc,taskName)
    dbk = copy.deepcopy(dbk_nullGuid); dbk.flight = 'flTAKEOFF'; OA.addDataBlocks(G,nc,dbk)

    nc = OA.GBL.FINALNODE;   taskName="finalEDT"; OA.graphAddNode(G,nc,taskName)
    dbk = copy.deepcopy(dbk_nullGuid); dbk.flight = 'flLANDING'; OA.addDataBlocks(G,nc,dbk)

    nc += 1;  taskName="beginAlgoEDT"; OA.graphAddNode(G,nc,taskName)
    dbk = copy.deepcopy(dbk_parabola); dbk.flight = 'flTAKEOFF';  dbk.localname='o_parabola'; OA.addDataBlocks(G,nc,dbk)
    OA.addCustomText(G, nc, 'err = setupTheParabola(o_parabola); IFEB;')

    nc += 1;  taskName="SolveEDT"; OA.graphAddNode(G,nc,taskName)
    dbk = copy.deepcopy(dbk_parabola); dbk.flight = 'flLANDING'; dbk.localname='in_parabola'; OA.addDataBlocks(G,nc,dbk)
    dbk = copy.deepcopy(dbk_roots); dbk.flight = 'flTAKEOFF'; dbk.localname='o_roots'; OA.addDataBlocks(G,nc,dbk)
    OA.addCustomText(G, nc, 'err = solveForRoots(in_parabola, o_roots); IFEB;')

    nc += 1;  taskName="CountEDT"; OA.graphAddNode(G,nc,taskName)
    dbk = copy.deepcopy(dbk_roots); dbk.flight = 'flTAGO'; dbk.localname='io_roots'; OA.addDataBlocks(G,nc,dbk)
    OA.addCustomText(G, nc, 'err = deduplicateRoots(io_roots); IFEB;')

    nc += 1;  taskName="PrintEDT"; OA.graphAddNode(G,nc,taskName)
    dbk = copy.deepcopy(dbk_roots); dbk.flight = 'flTAGO'; dbk.localname='io_roots'; OA.addDataBlocks(G,nc,dbk)
    OA.addCustomText(G, nc, 'err = printRoots(io_roots); IFEB;')

    nc += 1;  taskName="endAlgoEDT"; OA.graphAddNode(G,nc,taskName)
    dbk = copy.deepcopy(dbk_roots); dbk.flight = 'flLANDING'; OA.addDataBlocks(G,nc,dbk)

    # ----- EDGES
    edg = OA.graphAddEdge(G,     "mainEdt", "beginAlgoEDT", "NULL_GUID")
    OA.getEvent(G, edg).eflag = 'EVT_PROP_NONE'
    OA.getEvent(G, edg).satisfy = 'NULL_GUID'

    edg=OA.graphAddEdge(G, "beginAlgoEDT", "SolveEDT", "Parabola")
    OA.getEvent(G, edg).accessMode = 'DB_MODE_EW'
    OA.getEvent(G, edg).satisfy = OA.makeGuidDataBlockname('Parabola')

    edg=OA.graphAddEdge(G,    "SolveEDT",     "CountEDT", "Roots")
    OA.getEvent(G, edg).accessmode = 'DB_MODE_EW'
    OA.getEvent(G, edg).satisfy = OA.makeGuidDataBlockname('Roots')

    edg=OA.graphAddEdge(G,    "CountEDT",     "PrintEDT", "Roots")
    OA.getEvent(G, edg).accessMode = 'DB_MODE_CONST'
    OA.getEvent(G, edg).satisfy = OA.contextedGuid(OA.GBL.DBK_structName +'->', OA.makeGuidDataBlockname('Roots'), 'DBK')

    edg=OA.graphAddEdge(G,    "PrintEDT",   "endAlgoEDT", "Roots")
    OA.getEvent(G, edg).accessMode = 'DB_MODE_CONST'
    OA.getEvent(G, edg).satisfy = OA.contextedGuid(OA.GBL.DBK_structName +'->', OA.makeGuidDataBlockname('Roots'), 'DBK')

    edg=OA.graphAddEdge(G,  "endAlgoEDT",     "finalEDT", "NULL_GUID")
    OA.getEvent(G, edg).eflag = 'EVT_PROP_NONE'
    OA.getEvent(G, edg).satisfy = 'NULL_GUID'

    # ----- WORK
    erri = 0
    while not erri:
        if 1 == 2: OA.printGraph(G)   # Just to check if the inputs are ok.

        erri = OA.OCRanalysis(G)
        if erri: break

        if 1 == 2: OA.printGraph(G)
        if 1 == 2:
            erri = OA.outputDot(G, "z_helloWorld.dot")
            if erri: break
        if 1 == 1:
            erri = OA.outputOCR(G, "z_helloWorld.c")
            if erri: break

        if 1 == 2: OA.printGraph(G)

        break  # while not erri
    OA.errmsg(erri)
    return erri
# ------------------------------------------------------------------------------
# ------------------------------------------------------------------------------
if __name__ == "__main__":
    errm = theMain()
    if errm:
        exit(1)
    else:
        exit(0)

